README.md
[![Build Status](https://app.travis-ci.com/t2ym/i18n-number.svg?branch=master)](https://app.travis-ci.com/t2ym/i18n-number)
[![Coverage Status](https://coveralls.io/repos/github/t2ym/i18n-number/badge.svg?branch=master&build=26)](https://coveralls.io/github/t2ym/i18n-number?branch=master)
[![npm version](https://badge.fury.io/js/i18n-number.svg)](https://badge.fury.io/js/i18n-number)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/t2ym/i18n-number)

### i18n-number

Wrapper element for [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat).

[Demo](https://t2ym.github.io/i18n-number/components/i18n-number/demo/) and [API Docs](https://www.webcomponents.org/element/t2ym/i18n-number/elements/i18n-number)

<img src="https://raw.githubusercontent.com/wiki/t2ym/i18n-number/i18n-number-demo.gif" width="600px">

### Install

```
    npm install i18n-number
```

### Import

```js
    import "i18n-number/i18n-number.js";
```

### Usage

```html
    <i18n-number 
      lang="en"
      options='{ "style": "currency", "currency": "USD" }' 
      >123456.78</i18n-number>
```

This renders as follows:

```
    $123,456.78
```

### Demo

#### Prerequisite for Building and Serving Demo

```sh
    npm install -g polymer-cli
```

#### On-the-fly Build

```sh
    # Serve at http://localhost:8080/components/i18n-number/demo/
    polymer serve --npm --module-resolution=node -p 8080
```

#### Static Builds

- Build

```sh
    polymer build
```

- Targets

| target | browser | minify | bundled |
|:-----:|:-------:|:------:|:-------:|
|build/esm-unbundled|es6,modules|no|no|
|build/esm-bundled  |es6,modules|yes|yes|
|build/es6-bundled  |es6        |yes|yes|
|build/es5-bundled  |es5        |yes|yes|

- Serve

```sh
    # Serve at http://localhost:8080/
    cd build/{esm-unbundled|esm-bundled|es6-bundled|es5-bundled}
    python -m SimpleHTTPServer 8080 # or any HTTP(S) server
```

### License

[BSD-2-Clause](https://github.com/t2ym/i18n-number/blob/master/LICENSE.md)


CHANGELOG.md
# Changelog

## [5.0.0] 2021-9-28
### Added
- Support lit-html@2.0.0

### Changed

### Removed

## [4.0.1] 2020-3-27
### Added

### Changed
- Update package-lock.json
- Update test environments for Travis CI

### Removed
- Support for rare locales like zh-yue-Hans-CN, zh-CHS
- test: Remove Travis CI tests on SauceLabs to avoid connection refused errors
  - Tests pass from local to SauceLabs

## [4.0.0] 2019-3-20
### Added
- Depending on `lit-html@^1.0.0`
- Depending on `wc-putty@^0.1.0`

### Changed
- Rendering with `lit-html@^1.0.0` instead of Polymer library

### Removed
- Dependency on `@polymer/polymer`
- Safari 9 support

## [3.0.0] 2019-2-24
### Added
- Support Polymer 3.x in ES modules

### Changed

### Removed
- Polymer 1.x/2.x hybrid support in HTML Imports
- Bower support

## [2.0.0]
### Added
- Hybrid support of Polymer 1.x/2.x in the legacy syntax

### Changed

## [1.0.1]
### Added
- Polymer 1.x support

### Changed

### Removed


test/bound-i18n-number.js
/**
@license https://github.com/t2ym/i18n-number/blob/master/LICENSE.md
Copyright (c) 2016, Tetsuya Mori <t2y3141592@gmail.com>. All rights reserved.
*/
import { Polymer, html } from '@polymer/polymer/polymer-legacy.js';
import '@polymer/iron-selector/iron-selector.js';
Polymer({
  importMeta: import.meta,

  _template: html`
    <span id="placeholder">
      <i18n-number id="num">{{number}}</i18n-number>
      <i18n-number id="num2">{{number2}}</i18n-number>
      <i18n-number id="empty"></i18n-number>
      <i18n-number id="currency" options="{{options}}">123456.789</i18n-number>
      <iron-selector id="currencySelector" attr-for-selected="value" selected="{{options.currency}}">
        <div name="USD" iron-selected="">USD</div>
        <div name="JPY">JPY</div>
      </iron-selector>
    </span>
`,

  is: 'bound-i18n-number',

  properties: {
    number: {
      type: String,
      value: '',
      notify: true
    },
    number2: {
      type: String,
      value: '12.34',
      notify: true
    },
    options: {
      type: Object,
      value: function () {
        return { 'style': 'currency', 'currency': 'USD' };
      },
      notify: true
    }
  },

  attached: function() {
    this.$.num.setAttribute('lang', this.$.num.getAttribute('lang'));
    this.$.num.setAttribute('lang', this.$.num.getAttribute('lang'));
    this.$.num.setAttribute('attr', 'value');
    this.$.num.setAttribute('options', '<');
    this.$.num.constructor.importMeta;
    this.$.num.root;
  }
});


test/selenium-grid-wct.conf.js
module.exports = {
  "webserver": {
    "hostname": "selenium-hub"
  },
  "activeBrowsers": [
    {
      "url": 'http://selenium-hub:4444/wd/hub',
      "browserName": "internet explorer",
      "version": "11",
    },
    {
      "url": 'http://selenium-hub:4444/wd/hub',
      "browserName": "MicrosoftEdge",
      "version": "17",
    },
    {
      "url": 'http://selenium-hub:4444/wd/hub',
      "browserName": "safari",
      "version": "12",
    },
  ],
  "plugins": {
    "local": {
      "browsers": [
        "chrome",
        "firefox"
      ],
      "browserOptions": {
        "chrome": [
          "headless",
          "disable-gpu"
        ],
        "firefox": [
          "--headless"
        ]
      }
    },
    "sauce": {
      "disabled": true,
      "browsers": [
        {
          "browserName":  "microsoftedge",
          "platform":     "Windows 10",
          "version":      "17"
        },
        {
          "browserName":  "internet explorer",
          "platform":     "Windows 8.1",
          "version":      "11"
        },
        {
          "browserName":  "safari",
          "platform":     "OS X 10.13",
          "version":      "12"
        },
        {
          "browserName":  "safari",
          "platform":     "OS X 10.11",
          "version":      "10"
        }
      ]
    },
    "istanbul": {
      "dir": process.env.COVERAGE_DIR || "test/coverage-report",
      "reporters": ["text-summary", "lcov"],
      "include": [
        "**/i18n-number.js"
      ],
      "exclude": [
        "/test/**/*"
      ]
    }
  }
};


i18n-number.js
/**
@license https://github.com/t2ym/i18n-number/blob/master/LICENSE.md
Copyright (c) 2016, Tetsuya Mori <t2y3141592@gmail.com>. All rights reserved.
*/
/**
`<i18n-number>` renders a number in international formats by Intl.NumberFormat 
(http://www.ecma-international.org/ecma-402/1.0/#sec-11.1) object.
If Intl.NumberFormat is unavailable, [Intl.js Polyfill](https://github.com/andyearnshaw/Intl.js) and 
its locale modules are dynamically loaded as a fallback.

    <i18n-number lang="en"
                 options='{ "style": "currency", "currency": "USD" }'
                 offset="1"
                 >123456.78</i18n-number>

This example renders the following number string.

    $123,455.78

@element i18n-number
@hero hero.svg
@demo demo/index.html
*/
import { polyfill } from 'wc-putty/polyfill.js';
import { html, render } from 'lit-html/lit-html.js';

const formatCache = new Map();

export class I18nNumber extends polyfill(HTMLElement) {
  static get is() {
    return 'i18n-number';
  }

  static get observedAttributes() {
    return [ 'lang', 'options', 'offset' ];
  }

  __render() {
    return html`<span id="number">${this.formatted || ''}</span>`;
  }

  /**
   * Fired whenever the formatted text is rendered.
   *
   * @event rendered
   */

  constructor() {
    super();
    /**
     * Default locale constant 'en'
     */
    this.DEFAULT_LANG = 'en';
    /**
     * The locale for the formatted number.
     * The typical value is bound to `{{effectiveLang}}` when the containing element has
     * `BehaviorsStore.I18nBehavior`.
     */
    //this.lang = this.DEFAULT_LANG; // put off to connectedCallback
    /**
     * Offset for number
     *
     * Note: number = rawNumber - offset 
     */
    this.offset = 0;
    /**
     * Raw string synchronized with textContent
     */
    //this.raw = undefined;
    /**
     * Options object for Intl.NumberFormat 
     * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat)
     */
    //this.options = undefined;
    /**
     * Raw number parsed from raw
     */
    //this.rawNumber = undefined;
    /**
     * Number calculated from rawNumber and offset
     */
    //this.number = undefined;
    /**
     * Formatted string rendered for UI
     */
    //this.formatted = undefined;
  }

  attributeChangedCallback(name, oldValue, newValue) {
    switch (name) {
    case 'lang':
      this._langChanged(newValue);
      break;
    case 'options':
      try {
        this.options = JSON.parse(newValue);
      }
      catch (ex) {
        this.options = undefined;
      }
      break;
    case 'offset':
      this.offset = parseInt(newValue);
      break;
    /* istanbul ignore next */
    default:
      /* istanbul ignore next */
      break;
    }
  }

  /**
   * root property to imitate a Polymer element
   */
  get root() {
    return this.shadowRoot;
  }

  /**
   * options property for options object for Intl.NumberFormat
   * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat)
   *
   * Note: _optionsChanged() is called on every change
   */
  get options() {
    return this._options;
  }

  set options(value) {
    this._optionsChanged(this._options = value);
  }

  /**
   * raw property to store a raw number string from textContent
   *
   * Note: _rawChanged() is called on every change
   */
  get raw() {
    return this._raw;
  }

  set raw(value) {
    this._rawChanged(this._raw = value);
  }

  /**
   * Offset for number
   *
   * Note:
   *  - Calculation: number = rawNumber - offset
   *  - _offsetChanged() is called on every change
   */
  get offset() {
    return this._offset;
  }

  set offset(value) {
    this._offsetChanged(this._offset = value);
  }

  /**
   * connectedCallback for custom elements
   *
   * Tasks:
   *  - Sets this.lang if not set
   *  - Sets up observers if not set up
   *  - Sets this.raw from this.textNode.data
   *  - Triggers rendering
   */
  connectedCallback() {
    if (!this.lang) {
      this.lang = this.DEFAULT_LANG;
    }
    if (!this.observer) {
      this._setupObservers();
    }
    this.raw = this.textNode.data;
    this.invalidate();
  }

  /**
   * Sets up observers of textContent mutations
   */
  _setupObservers() {
    let i = 0;
    do {
      this.textNode = this.childNodes[i++];
      if (!this.textNode) {
        this.textNode = this.childNodes[0];
        break;
      }
    }
    while (this.textNode.nodeType !== this.textNode.TEXT_NODE);
    if (!this.textNode) {
      this.appendChild(document.createTextNode(''));
      this.textNode = this.childNodes[0];
    }
    this.observer = new MutationObserver(this._textMutated.bind(this));
    this.observer.observe(this.textNode, { characterData: true });
    this.nodeObserver = new MutationObserver(function (mutations) {
      mutations.forEach(function(mutation) {
        switch (mutation.type) {
        case 'childList':
          let i = 0;
          do {
            if (mutation.addedNodes[i] &&
                mutation.addedNodes[i].nodeType === mutation.addedNodes[i].TEXT_NODE) {
              this.textNode = mutation.addedNodes[i];
              this.raw = this.textNode.data;
              //console.log('i18n-number: text node added with ' + this.raw);
              this.observer.observe(this.textNode, { characterData: true });
              break;
            }
            i++;
          }
          while (i < mutation.addedNodes.length);
          break;
        /* istanbul ignore next */
        default:
          /* istanbul ignore next: mutation.type is characterData or attributes */
          break;
        }
      }, this);
    }.bind(this));
    this.nodeObserver.observe(this, { childList: true });
  }

  /**
   * MutationObserver callback of the child text node to re-render on text mutations.
   *
   * @param {Array} mutations Array of MutationRecord (https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).
   */
  _textMutated(mutations) {
    mutations.forEach(function(mutation) {
      switch (mutation.type) {
      case 'characterData':
        //console.log('i18n-number: _textMutated: raw = ' + mutation.target.data);
        if (this.raw !== mutation.target.data) {
          this.raw = mutation.target.data;
        }
        break;
      /* istanbul ignore next: mutation.type is characterData */
      default:
        /* istanbul ignore next: mutation.type is characterData */
        break;
      }
    }, this);
  }

  /**
   * Observer of `raw` property to re-render the formatted number.
   *
   * @param {string} raw New raw number string.
   */
  _rawChanged(raw) {
    if (this.textNode) {
      if (raw !== this.textNode.data) {
        this.textNode.data = raw;
      }
      //console.log('i18n-number: _rawChanged: raw = ' + raw);
      this._render(this.lang, this.options, raw, this.offset);
    }
  }

  /**
   * Observer of `lang` property to re-render the formatted number.
   *
   * @param {string} lang New locale.
   */
  _langChanged(lang) {
    if (!lang || lang === 'null') {
      this.lang = this.DEFAULT_LANG;
      lang = this.lang;
    }
    if (this.textNode) {
      //console.log('i18n-number: _langChanged: lang = ' + lang);
      this._render(lang, this.options, this.raw, this.offset);
    }
  }

  /**
   * Observer of `options` property to re-render the formatted number.
   *
   * @param {Object} options New options for Intl.NumberFormat.
   */
  _optionsChanged(options) {
    if (this.textNode) {
      //console.log('i18n-number: _optionsChanged: options = ' + JSON.stringify(options));
      this._render(this.lang, options, this.raw, this.offset);
    }
  }

  /**
   * Partially emulates notifyPath() in Polymer library
   * Just calls _onOptionsPropertyChanged() to re-render
   */
  notifyPath(path, value) {
    this._onOptionsPropertyChanged();
  }

  /**
   * Observer of `options` sub-properties to re-render the formatted number.
   */
  _onOptionsPropertyChanged(/* changeRecord */) {
    if (this.textNode) {
      //console.log('_onOptionsPropertyChanged: path = ' + changeRecord.path + ' value = ' + JSON.stringify(changeRecord.value));
      this._render(this.lang, this.options, this.raw, this.offset);
    }
  }

  /**
   * Observer of `offset` property to re-render the formatted number.
   *
   * @param {number} offset New offset.
   */
  _offsetChanged(offset) {
    if (this.textNode) {
      //console.log('i18n-number: _offsetChanged: offset = ' + offset);
      this._render(this.lang, this.options, this.raw, offset);
    }
  }

  /**
   * Gets a cached Intl.NumberFormat object
   *
   * @param {string} lang Locale for formatting.
   * @param {Object} options Options for Intl.NumberFormat.
   * @return {Object} Intl.NumberFormat object.
   */
  _getNumberFormatObject(lang, options) {
    if (!lang || lang === 'null') {
      lang = this.DEFAULT_LANG;
    }
    let formatId = lang + JSON.stringify(options);
    let formatObject = formatCache.get(formatId);
    if (!formatObject) {
      formatObject = new Intl.NumberFormat(lang, options);
      formatCache.set(formatId, formatObject);
    }
    return formatObject;
  }

  /**
   * Formats the number
   *
   * @param {string} lang Locale for formatting.
   * @param {Object} options Options for Intl.NumberFormat.
   * @param {number} number Number to format.
   * @return {string} Formatted number string.
   */
  _formatNumber(lang, options, number) {
    if (!lang) {
      lang = this.DEFAULT_LANG;
    }
    try {
      return this._getNumberFormatObject(lang, options).format(number);
    }
    catch (e) {
      return number.toString();
    }
  }

  /**
   * Renders the formatted number
   *
   * @param {string} lang Locale for formatting.
   * @param {Object} options Options for Intl.NumberFormat.
   * @param {string} raw Raw number string.
   * @param {number} offset Offset for number.
   */
  _render(lang, options, raw, offset) {
    // TODO: rendering may be done redundantly on property initializations
    raw = raw.trim();
    if (!raw && !this.formatted) {
      //console.log('i18n-number: skipping _render as raw is null');
      return;
    }
    if (raw) {
      this.rawNumber = Number(raw);
      this.number = this.rawNumber - offset;
      this.formatted = this._formatNumber(lang, options, this.number);
    }
    else {
      this.rawNumber = undefined;
      this.number = undefined;
      this.formatted = '';
    }
    this.invalidate();
    //console.log('i18n-number: _render ' + this.formatted);
  }

  /**
   * Renders shadowRoot with this.__render()
   */
  invalidate() {
    if (!this.needsRender) {
      this.needsRender = true;
      Promise.resolve().then(() => {
        this.needsRender = false;
        if (!this.shadowRoot) {
          this.attachShadow({ mode: 'open' });
        }
        render(this.__render(), this.shadowRoot);
        //console.log(`rendered "${this.formatted}"`);
        if (typeof this.formatted !== 'undefined') {
          this.dispatchEvent(new Event('rendered', { bubbles: true, cancelable: false, composed: true }));
        }
      });
    }
  }

  /**
   * Renders the formatted number with the current parameters
   *
   * Note: (As of Polymer 1.2.3)
   *   Explicit render() call is needed whenever the observer 
   *   `_onOptionsPropertyChanged(options.*)` is NOT invoked 
   *   after a property of `options` is changed.  An explicit call 
   *   `this.notifyPath('options', this.options, true)` can also 
   *   trigger re-rendering.
   *
   *   If the changed property of `options` is bound in an annotation
   *   like `{{options.currency}}`, the observer `_onOptionsPropertyChanged(options.*)`
   *   is automatically called whenever the property value is changed
   *   and thus no explicit call of `render()` or `notifyPath()` is
   *   required.
   */
  render() {
    this._render(this.lang, this.options, this.raw, this.offset);
  }
}
customElements.define(I18nNumber.is, I18nNumber);


demo/index.js
import "@polymer/iron-flex-layout/iron-flex-layout-classes.js";
import "@polymer/paper-input/paper-input.js";
import "@polymer/paper-dropdown-menu/paper-dropdown-menu.js";
import "@polymer/paper-item/paper-item.js";
import "@polymer/paper-styles/demo-pages.js";
import "@polymer/iron-demo-helpers/demo-snippet.js";
import "@polymer/iron-demo-helpers/demo-pages-shared-styles.js";
import "../i18n-number.js";
import "./nano-calc.js";

{
  const $_documentContainer = document.createElement('template');
  $_documentContainer.setAttribute('style', 'display: none;');

  $_documentContainer.innerHTML = `<style is="custom-style" include="demo-pages-shared-styles">
        .vertical-section-container {
          max-width: 768px;
        }

        @media (max-width: 640px) {

          body {
            font-size: 14px;
            margin: 0;
            padding: 4px;
            background-color: var(--paper-grey-50);
          }

          .horizontal-section {
            background-color: white;
            padding: 8px;
            margin-right: 4px;
            min-width: 200px;

            @apply(--shadow-elevation-2dp);
          }

          .vertical-section {
            background-color: white;
            padding: 8px;
            margin: 0 4px 8px 4px;

            @apply(--shadow-elevation-2dp);
          }

        }
      </style>`;

  document.head.appendChild($_documentContainer.content);
}
{
  const $_documentContainer = document.createElement('template');
  $_documentContainer.setAttribute('style', 'display: none;');

  $_documentContainer.innerHTML = `<style is="custom-style">
        --demo-snippet-code {
          @apply(--paper-font-code1);
        }
    
        paper-item {
          display: block;
        }

        code {
          font-family: 'Roboto Mono', 'Consolas', 'Menlo', monospace;
          color: black;
        }

        .input {
          font-family: 'Roboto', sans-serif;
          font-size: 16px;
          border-width: 0px;
          outline-width: 0px;
        }

        .select {
          font-family: 'Roboto', sans-serif;
          font-size: 16px;
          outline-width: 0px;
          border-width: 0px;
          background-color: rgba(255,255,255,0.0);
        }

        option {
          position: relative;
          top: -1;
          color: black;
          outline-width: 0px;
          border-width: 0px;
        }

        .dropdown {
          --editable-dropdown-menu: {
            display: inline-block;
            margin-right: 8px;
            text-align: left;
            width: 180px;
          };
        }

        .demo-paper-dropdown-menu {
          font-family: 'Roboto', 'Noto', Helvetica, sans-serif;
          font-size: 16px;
          font-weight: 400;
          line-height: 24px;
          text-align: left;
          margin: auto;
          width: 180px;
        }

        @media (max-width: 640px) {

          body {
            font-size: 14px;
            margin: 0;
            padding: 4px;
            background-color: var(--paper-grey-50);
          }

          .horizontal-section {
            background-color: white;
            padding: 8px;
            margin-right: 4px;
            min-width: 200px;

            @apply(--shadow-elevation-2dp);
          }

          .vertical-section {
            background-color: white;
            padding: 8px;
            margin: 0 4px 8px 4px;

            @apply(--shadow-elevation-2dp);
          }

        }
      </style>`;

  document.head.appendChild($_documentContainer.content);
}
{
  const $_documentContainer = document.createElement('template');

  $_documentContainer.innerHTML = `<div class="vertical-section-container centered">

        <h1><code>&lt;i18n-number&gt;</code> Demo</h1>

        <div class="vertical-section-container">
          <h2>Number in English (en) locale</h2>
          <demo-snippet class="centered-demo">
            <template>
              <i18n-number lang="en">123456.78</i18n-number>
            </template>
            <script type="text/markdown">
              <i18n-number lang="en">123456.78</i18n-number>
            <${''}/script>
          </demo-snippet>

          <h2>Currency - USD in English (en) locale</h2>
          <demo-snippet class="centered-demo">
            <template>
              <i18n-number lang="en" options="{ &quot;style&quot;: &quot;currency&quot;, &quot;currency&quot;: &quot;USD&quot; }">123456.78</i18n-number>
            </template>
            <script type="text/markdown">
              <i18n-number 
                lang="en"
                options='{ "style": "currency", "currency": "USD" }' 
                >123456.78</i18n-number>
            <${''}/script>
          </demo-snippet>

          <!--
          <h2><code>&lt;i18n-number&gt;</code> with different parameters</h2>
          <demo-snippet id="demo-snippet" class="left-aligned-demo" _markdown="{{markdown}}">
            <template>
              <i18n-number-demo id="i18n-number-demo"></i18n-number-demo>
            </template>
          </demo-snippet>
          -->

          <h2>Digits display for <code>&lt;nano-calc&gt;</code></h2>
          <demo-snippet class="centered-demo">
            <template>
              <nano-calc x="123456.789"></nano-calc>
            </template>
            <script type="text/markdown">
              <dom-module id="nano-calc">
                <template>
                  ...
                  <span class="layout horizontal digits-display">
                    <span class="flex"></span>
                    <i18n-number id="value" 
                      formatted="{{value}}"
                      options='{ "maximumFractionDigits": 16 }'
                      >{{pickDisplayValue(x,y,op)}}</i18n-number>
                    </span>
                  </span>
                  ...
                </template>
                ...
              </dom-module> 
            <${''}/script>
          </demo-snippet>


    

  </div></div>`;

  document.body.appendChild($_documentContainer.content);
}
